### PR TITLE
[WIP] _nested_view_from_buffer.cont, torch.cat([NTs], dim=0)

### PR DESCRIPTION
**Pull request by cpuhrsch**
TODO
- Tests

### CODE REVIEW

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 1]
```diff
@@ -236,6 +236,22 @@ NestedTensorImpl::NestedTensorImpl(
...
+          impl_type,
+          base_tensor,
+          nested_sizes,
+          construct_nested_strides(nested_sizes),
+          construct_offsets(nested_sizes)) {
+  validate_nested_tensor_metadata(
+      nested_sizes_, nested_strides_, storage_offsets_);
+  refresh_dim();
+  set_custom_sizes_strides(c10::TensorImpl::SizesStridesPolicy::CustomSizes);
+}
```
Comment (jbschlosser) [inline aten/src/ATen/NestedTensorImpl.cpp:253]: Not possible to call into another constructor and avoid the duplication?

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -154,7 +154,7 @@
     "fill.Scalar",  # only used by the functionalization pass
     "lift.*",
     "normal_functional",  # only used by the functionalization pas
-    "_nested_view_from_buffer",  # View only version of _nested_from_buffer. This will force users to only use the "safe" version.
+    # "_nested_view_from_buffer",  # View only version of _nested_from_buffer. This will force users to only use the "safe" version.
```
Comment (jbschlosser) [inline tools/autograd/gen_python_functions.py:157]: Reminder to remove this

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1286,6 +1286,7 @@
   dispatch:
     SparseCPU, SparseCUDA: cat_sparse
     QuantizedCPU: cat_quantized_cpu
+    NestedTensorCPU, NestedTensorCUDA: cat_nested
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:1289]: If we land this, I guess it'd be better to bump cat support to a separate PR to keep them modular

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 4]
```diff
@@ -914,5 +939,40 @@ Tensor& normal_nested_(Tensor& self, double mean, double std, c10::optional<Gene
...
+    buffers.push_back(t_ptr->get_buffer().view({-1}));
+    sizes.push_back(t_ptr->get_nested_sizes());
+  }
+  return at::detail::make_tensor<NestedTensorImpl>(
+      at::cat(buffers).view({-1}), at::cat(sizes, 0));
+}
+
+Tensor cat_nested(const ITensorListRef& tensors, int64_t dim) {
+  auto materialized = tensors.materialize();
+  auto maybe_outnames = namedinference::compute_cat_outnames(materialized);
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:970]: we still mess with named tensors?

**Review by jbschlosser — COMMENTED** [review]
`cat` support and contiguous constructor look correct. Added some logistics comments if we were to actually land this, but it seems good enough to play around with benchmarking. In particular, avoiding building contiguous strides / offsets in Python seems a good thing to do :)
